### ISSUE TITLE
Crunchy Output

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2775]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tehminkeh reported|
|-|-|
|Trac ID|trac#2775|
|Type|enhancement|
|Created|2007-08-06 18:56:04Z|

Adds a new output plugin for crunchy. Fixes pre tags because that is the only  way crunchy output differs from html output.

Attachments:

* [crunchy_output.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dce/dce17c17ddf7d39e3ea366deb2a6c692fa3b72d7/2deba69fa483a7609a068516895397f7b4e563ff.diff) (4724 bytes) - added by tehminkeh on 2007-08-06 18:56:17Z - 
* [crunchy_output.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dce/dce17c17ddf7d39e3ea366deb2a6c692fa3b72d7/2ec5a757858e17effe4584ab81d5a566cbd6cf2b.diff) (6866 bytes) - added by tehminkeh on 2007-08-06 20:32:55Z - 
* [crunchy_output.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dce/dce17c17ddf7d39e3ea366deb2a6c692fa3b72d7/970f75d64cb575a271fc0e06905c874184f5a131.diff) (7350 bytes) - added by tehminkeh on 2007-08-07 19:14:45Z - formatting and docstring fixes
* [crunchy_output.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dce/dce17c17ddf7d39e3ea366deb2a6c692fa3b72d7/371d5ff56fec55b54713fce7a89da33e63738228.diff) (8282 bytes) - added by tehminkeh on 2007-08-08 16:15:08Z - 
* [crunchy_output.5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dce/dce17c17ddf7d39e3ea366deb2a6c692fa3b72d7/c04e9c7dafd46d6a93cef6089c4fe682f9069398.diff) (8586 bytes) - added by tehminkeh on 2007-08-08 18:35:12Z - I now explain why we have this huge extra class

<details><summary>Searchable metadata</summary>

```
trac-id__2775 2775
type__enhancement enhancement
reporter__tehminkeh tehminkeh
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__lore lore
keywords__EndOfLore EndOfLore
time__1186426564000000 1186426564000000
changetime__1389978217000000 1389978217000000
version__None None
owner__ 
cc__therve cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to tehminkeh|
|-|-|

Let's try to go with an argument-passing strategy rather than using the pythonPrettifier global.

doFile is the entry-point for tree.py.  generate_html and generate_crunchy each invoke it (though generate_crunchy does so only indirectly at the moment).  doFile could gain a new parameter, the formatter, and each of the generate_ methods could pass an instance of a different class in.  doFile would need to pass this object down to whichever other function actually does the formatting, of course.

There are some other minor issues with the patch, generate_crunchy has no docstring, pythonCodePrettify should be new-style (subclass object) and follow the twisted naming convention (PythonCodePrettify).

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tehminkeh removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to tehminkeh|
|-|-|

This is looking pretty good.  There are some pretty trivial things I'd like changed first, then I think this can be applied:

  * remove all the tabs
  * Underscores are reserved for separated static parts of a name with parts which are dispatched on dynamically, so PythonCodePrettify_crunchy doesn't quite fit.  I'd suggest a name like CrunchyCodePrettifier.
  * Most of the code in each of the two implementations of fontifyPythonNode is the same.  This is particularly unfortunate since the code in question is so bad. ;)  I'd suggest factoring most of it into a method on a base class and then having two differently behaving subclasses which call it.  (Alternatively, PythonCodePrettify might be the base class and have two methods, and the crunch version might subclass it and just override one of them).  Having a method which does, perhaps, everything except for the replaceChild call, and returns the new Node should let you have two very trivial other methods which just call it, one of which then does only a replaceChild and the other of which restores the attributes and then does the replaceChild.
  * each function which takes a pythonPrettifier should document that parameter
  * a note about where more information about crunchy can be found might be appropriate in the generate_crunchy docstring

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tehminkeh set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to tehminkeh|
|-|-|

I suggest doing ```fontifyPythonNode = HTMLPythonPrettify().fontifyPythonNode``` in tree, and make it the default of subsequent calls (```def fontifyPython(document, pythonPrettifier=fontifyPythonNode):```, same in munge and doFile). The fact that it became a class is useless. Otherwise, that looks good.

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tehminkeh set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Currently blocked on [#2526](https://github.com/twisted/twisted/issues/2526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @khorn|
|-|-|

khorn is doing stuff with documentation.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_11"></a>|@khorn set status to `closed`|
|-|-|

Lore is no longer being used for Twisted documentation, and is being deprecated, prior to eventual removal, per [#6907](https://github.com/twisted/twisted/issues/6907).

This issue will not be fixed.
